### PR TITLE
Add callback kind to solc

### PR DESCRIPTION
**Pull request by leonardoalt**
Depends on #410

New version of https://github.com/ethereum/solc-js/pull/316

Depends on https://github.com/ethereum/solidity/pull/7447

With this change, one can call
`compileStandard(..., smtchecker.smtCallback(smtsolver.smtSolver), ...)`
and get the SMTChecker working only requiring a local `z3` or `cvc4`.

### CODE REVIEW

**Comment by leonardoalt** [comment]
Still needs tests.

File: wrapper.js [hunk 1]
```diff
@@ -26,7 +26,7 @@ function setupMethods (soljson) {
   };
 
   // This calls compile() with args || cb
-  var runWithReadCallback = function (readCallback, compile, args) {
+  var runWithReadCallback = function (readCallback, auxInputCallback, compile, args) {
```
Comment (leonardoalt) [inline wrapper.js]: @axic new argument here or at the end?

**Comment by leonardoalt** [comment]
Do we want to give this callback by default if none is provided?

**Comment by leonardoalt** [comment]
Current state:
If a function is given to `runWithCallbacks` it uses it as the old `readCallback`, so nothing changes there.
If an object `callbacks` is given to `runWithCallbacks`, use `callbacks.importCallback` as the old `readCallback` and `callbacks.smtSolverCallback` as the new SMT callback.

In any case, they are wrapped into a single callback that decodes the required callback type. This is the breaking change in the data. Solidity will send the callback type before the data:
- import callback `sources:`
- SMT solver callback `smt-query:`

**Comment by leonardoalt** [comment]
@axic @chriseth

**Comment by leonardoalt** [comment]
The tests added here require
- `z3` or `cvc4` installed locally
- `smtCheckerTests` in `test`
Do we want the `solc-js` tests do take care of these, or should the Solidity solc-js tests do that?

**Comment by leonardoalt** [comment]
Tentative decisions from call:
- Instead of encoding the callback kind as a prefix in the data, break the C interface and add a new parameter for the callback kind.
- The above implies we need to check the version to decide the number of arguments for `solidity_compile` from `soljson`.
- Break the solc-js `compileStandardWrapper`: only take an object of callbacks instead of `function || object`.
- solc-js runs simple SMT tests just to test interface, can install solver in CI just for tests
- Solidity runs the `smtCheckerTests` with `SMTLib2Interface` by copying `smtCheckerTests` to `solc-js/test/` and running the solc-js tests.

**Comment by leonardoalt** [comment]
The tests here will only pass with `soljson.js` from https://github.com/ethereum/solidity/pull/7447

File: wrapper.js [hunk 2]
```diff
@@ -14,8 +14,8 @@ function setupMethods (soljson) {
 
   var wrapCallback = function (callback) {
     assert(typeof callback === 'function', 'Invalid callback specified.');
-    return function (path, contents, error) {
-      var result = callback(soljson.Pointer_stringify(path));
+    return function (kind, data, contents, error) {
+      var result = callback(soljson.Pointer_stringify(kind), soljson.Pointer_stringify(data));
```
Comment (axic) [inline wrapper.js]: Will we add the context too? In a separate PR?

File: smtsolver.js [hunk 3]
```diff
@@ -53,5 +53,6 @@ function solve (query) {
 }
 
 module.exports = {
-  smtSolver: solve
+  smtSolver: solve,
+  availableSolvers: solvers
```
Comment (axic) [inline smtsolver.js]: Why is this needed? And if it is needed, would it make sense mapping it into names only?

File: wrapper.js [hunk 4]
```diff
@@ -26,20 +26,48 @@ function setupMethods (soljson) {
   };
 
   // This calls compile() with args || cb
-  var runWithReadCallback = function (readCallback, compile, args) {
+  var runWithCallbacks = function (callbacks, compile, args) {
+    var readCallback;
+    var smtSolverCallback;
+
+    if (typeof callbacks === 'function') {
```
Comment (axic) [inline wrapper.js]: Did we decide to not have this workaround in 0.6 then because we allow people to upgrade to the new syntax in 0.5?

File: wrapper.js (hunk 2 above)
Comment (leonardoalt) [reply wrapper.js]: Yeah I guess separate PR is better

File: smtsolver.js (hunk 3 above)
Comment (leonardoalt) [reply smtsolver.js]: It is needed by the `smtcallback.js` test such that it can decide not to run the tests if there's no solver installed. Names only should be fine. Even just the amount would be fine

File: wrapper.js (hunk 4 above)
Comment (leonardoalt) [reply wrapper.js]: Yea need to rebase and will remove this

File: smtsolver.js (hunk 3 above)
Comment (leonardoalt) [reply smtsolver.js]: Updated

**Comment by leonardoalt** [comment]
@axic This is ready for review, but the test that fails is not going to work without `soljson.js` from Solidity `develop_060`

File: test/smtcallback.js [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+          sources: test.solidity
+        }),
+        { smtSolverCallback: smtchecker.smtCallback(smtsolver.smtSolver) }
+      ));
+      st.ok(output);
+
+      // Collect obtained error messages
+      test.errors = collectErrors(output);
+
+      // Compare expected vs obtained errors
```
Comment (erak) [inline test/smtcallback.js]: Is this supposed to be in line 150?

**Comment by axic** [comment]
Why is this still done on `master_060` (and/or why is `master_060` out of date compared to `master`)?

**Comment by leonardoalt** [comment]
Aren't we following the same idea from Solidity here, to keep a separate 060 branch? We should merge master into master_060 once in a while, as it's done there.

File: test/smtcallback.js (hunk 5 above)
Comment (leonardoalt) [reply test/smtcallback.js]: Right

**Comment by leonardoalt** [comment]
Updated @erak 's comment.
@axic Anything else needed here?

**Comment by axic** [comment]
It is failing tests? ("Invalid callback kind specified")

**Comment by leonardoalt** [comment]
Yes, this is because it uses `soljson.js` from the latest release.
I think the merge order should be:
1) Merge https://github.com/ethereum/solidity/pull/7447 into Solidity `develop_060`. The external tests still do not pass there but I think I can write a tiny `solc-js` PR that will make it pass.
2) Change this PR to use `soljson.js` from `develop_060`. This would make the tests here pass as well.

**Comment by leonardoalt** [comment]
Actually we can't make the tests here pass without making it download `soljson.js` from `develop_060`.

File: wrapper.js [hunk 6]
```diff
@@ -26,25 +26,46 @@ function setupMethods (soljson) {
...
-  var runWithReadCallback = function (readCallback, compile, args) {
-    // Forward compatibility with 0.6.x
-    if (typeof readCallback === 'object') {
-      readCallback = readCallback.import;
+  var runWithCallbacks = function (callbacks, compile, args) {
+    var readCallback;
+    var smtSolverCallback;
+
+    if (callbacks !== undefined) {
+      readCallback = callbacks.importCallback;
```
Comment (axic) [inline wrapper.js]: Actually we called this `import` in 0.5.x.

File: wrapper.js [hunk 7]
```diff
@@ -71,21 +92,21 @@ function setupMethods (soljson) {
...
+      return runWithCallbacks(readCallback, compileInternal, [ input, optimize ]);
     };
   }
 
   var compileStandard = null;
   if ('_compileStandard' in soljson) {
     var compileStandardInternal = soljson.cwrap('compileStandard', 'string', ['string', 'number']);
     compileStandard = function (input, readCallback) {
-      return runWithReadCallback(readCallback, compileStandardInternal, [ input ]);
+      return runWithCallbacks(readCallback, compileStandardInternal, [ input ]);
```
Comment (axic) [inline wrapper.js]: Hm, this will, fail, won't it?

File: smtsolver.js [hunk 8]
```diff
@@ -3,14 +3,16 @@ var execSync = require('child_process').execSync;
...
 var potentialSolvers = [
   {
     name: 'z3',
-    params: '-smt2'
+    params: '-smt2 -t:' + timeout
   },
   {
     name: 'cvc4',
-    params: '--lang=smt2'
+    params: '--lang=smt2 --tlimit=' + timeout
```
Comment (axic) [inline smtsolver.js]: If there is a plan for another 0.5 release, then this timeout stuff could be merged into `master` first.

Comment (leonardoalt) [reply smtsolver.js]: Right

File: wrapper.js (hunk 6 above)
Comment (leonardoalt) [reply wrapper.js]: Updated

File: wrapper.js (hunk 7 above)
Comment (axic) [reply wrapper.js]: Actually realised the PR omitted the check for version. As we agreed two different ways are needed to pass callbacks based on the compiler version.

Comment (leonardoalt) [reply wrapper.js]: True, this and the other 2 above it need the old `readWithReadCallback`

Comment (axic) [reply wrapper.js]: All of them need it. For these above you can hardcode the old interface and make it version dependent under `solidity_compile`.

Comment (leonardoalt) [reply wrapper.js]: Why not only these two? The others don't have the read callback.

Comment (axic) [reply wrapper.js]: I meant all of the ones using `runWithCallbacks`.

File: test/compiler.js [hunk 9]
```diff
@@ -4,6 +4,16 @@ const solc = require('../index.js');
 const linker = require('../linker.js');
 const execSync = require('child_process').execSync;
 
+function wrapImportCallbackVersion (solc, importCallback) {
+  var callback;
+  if (semver.gt(solc.semver(), '0.5.99')) {
+    callback = { import: importCallback };
+  } else {
+    callback = importCallback;
```
Comment (axic) [inline test/compiler.js]: Could just use `return` statements.

File: test/compiler.js [hunk 10]
```diff
@@ -462,15 +476,15 @@ function runTests (solc, versionText) {
...
+        var output = JSON.parse(solc.compile(JSON.stringify(input), wrapImportCallbackVersion(solc, findImports)));
         var x = getBytecodeStandard(output, 'cont.sol', 'x');
         st.ok(typeof x === 'string');
         st.ok(x.length > 0);
         var L = getBytecodeStandard(output, 'lib.sol', 'L');
         st.ok(typeof L === 'string');
         st.ok(L.length > 0);
 
-        var outputNewApi = JSON.parse(solc.compile(JSON.stringify(input), { import: findImports }));
+        var outputNewApi = JSON.parse(solc.compile(JSON.stringify(input), wrapImportCallbackVersion(solc, findImports)));
```
Comment (axic) [inline test/compiler.js]: These two (`solc.compile`) should not use the wrapper, since the goal of their interface is to hide this.

File: test/compiler.js [hunk 11]
```diff
@@ -224,7 +235,7 @@ function runTests (solc, versionText) {
           throw new Error('Could not implement this interface properly...');
         }
         st.throws(function () {
-          solc.lowlevel.compileCallback(JSON.stringify({sources: input}), 0, findImports);
+          solc.lowlevel.compileCallback(JSON.stringify({sources: input}), 0, wrapImportCallbackVersion(solc, findImports));
```
Comment (axic) [inline test/compiler.js]: 0.5 (and 0.6) don't have `solc.lowlevel.compileCallback`.

File: test/compiler.js [hunk 12]
```diff
@@ -387,7 +398,10 @@ function runTests (solc, versionText) {
           }
         }
 
-        var output = JSON.parse(solc.lowlevel.compileStandard(JSON.stringify(input), findImports));
+        var output = JSON.parse(solc.lowlevel.compileStandard(
```
Comment (axic) [inline test/compiler.js]: This is the only case which should be needed.

File: test/smtcallback.js [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+      sources: input
+    });
+    var tests = [
+      { cb: satCallback, expectations: ['Assertion violation happens here'] },
+      { cb: unsatCallback, expectations: [] },
+      { cb: errorCallback, expectations: ['BMC analysis was not possible'] }
+    ];
+    for (var i in tests) {
+      var test = tests[i];
+      var output = JSON.parse(solc.compileStandardWrapper(
```
Comment (axic) [inline test/smtcallback.js]: Use `solc.compile`. `compileStandard` and `compileStandardWrapper` are deprecated aliases (should have removed them in 0.5.2, but will do in 0.6)

File: test/smtcallback.js [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+        { smtSolverCallback: test.cb }
+      ));
+      var errors = collectErrors(output);
+      st.ok(expectErrors(errors, test.expectations));
+    }
+    st.end();
+  });
+
+  t.test('Solidity smtCheckerTests', function (st) {
+    var testdir = './test/smtCheckerTests/';
```
Comment (axic) [inline test/smtcallback.js]: I think we should add a smoke test into this directory, otherwise none of this will be run on CI and it will be harder to find bugs introduced.

File: test/compiler.js (hunk 10 above)
Comment (leonardoalt) [reply test/compiler.js]: Right

File: test/compiler.js (hunk 9 above)
Comment (leonardoalt) [reply test/compiler.js]: Updated

File: test/compiler.js (hunk 10 above)
Comment (leonardoalt) [reply test/compiler.js]: Ah, but the interface hides this based on the version of the used `soljson`. If you use the new `soljon`, this breaks without this version wrapper

File: test/smtcallback.js (hunk 14 above)
Comment (leonardoalt) [reply test/smtcallback.js]: added

File: test/compiler.js [hunk 15]
```diff
@@ -583,7 +587,6 @@ function runTests (solc, versionText) {
 
       t.test('compiling standard JSON (multiple sources on old compiler)', function (st) {
         var output = JSON.parse(solc.compile('{"language":"Solidity","sources":{"cont.sol":{"content":"import \\"lib.sol\\";"},"lib.sol":{"content":""}}}'));
-        console.log(output);
```
Comment (leonardoalt) [inline test/compiler.js]: @axic I'm not sure why this was there, was it intended?

Comment (axic) [reply test/compiler.js]: No, it was typo during merging.
